client/components/LearnerViewDirectiveSpec.js

```diff
@@ -0,0 +1,201 @@
+
+  // Please make sure the following three variables are the same as
+  // what we have in LearnerViewDirective
+  var SECONDS_TO_MILLISECONDS = 1000;
+  var DEFAULT_AUTOSAVE_SECONDS = 5;
+
+  var DELTA_MILLISECONDS = 100;
+
+  var AUTOSAVE_MILLISECONDS =
+    DEFAULT_AUTOSAVE_SECONDS * SECONDS_TO_MILLISECONDS;
+
+  var AUTOSAVE_REPEAT_RANGE = 20;
+
+  beforeEach(inject(function($compile, $rootScope, _QuestionDataService_) {
+    $scope = $rootScope.$new();
+
+    // The reason why we have to go through this trouble to get $scope
+    // is the controller is anonymous, thus, there is no easy way to do
+    // it. Recommend future refactor.
+    element = angular.element('<learner-view></learner-view>');
+    element = $compile(element)($scope);
+    $scope = element.isolateScope();
+    $rootScope.$digest();
+    $scope.$digest();
+
+    QuestionDataService = _QuestionDataService_;
+
+  }));
+
+  describe("resetCode", function() {
+
+    it('should reset code to default code', function() {
+      $scope.questionIds.forEach(function(questionId, index) {
+        var question = QuestionDataService.getQuestion(questionId);
+        var starterCode = question.getStarterCode(LANGUAGE);
+        $scope.currentQuestionIndex = index;
+        $scope.code = generateRandomChars(NUM_CHARS_CODE);
+        expect(angular.equals(starterCode, $scope.code)).toBe(false);
+        $scope.resetCode();
+        expect($scope.code).toEqual(starterCode);
+      });
+    });
+  });
+
+  describe("navigateToQuestion", function() {
+    it('should navigate to given question', function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        $scope.navigateToQuestion(i);
+        expect($scope.currentQuestionIndex).toBe(i);
+      }
+    });
+  });
+
+  describe("activateAutosaving", function() {
+    var $interval;
+    var $timeout;
+    var CodeStorageService;
+    beforeEach(inject(function(_$interval_, _$timeout_, _CodeStorageService_) {
+      $interval = _$interval_;
+      $timeout = _$timeout_;
+      CodeStorageService = _CodeStorageService_;
+      localStorage.clear();
+    }));
+
+    // Autosave text is using $timeout, thus,
+    // we need to flush both $interval and $timeout.
+    var timeFlies = function(timeToFly) {
+      $interval.flush(timeToFly);
+      $timeout.flush(timeToFly);
+    };
+
+    it("should activate autosaving for only once", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        if (i !== 0) {
+          // Default question index is 0, we don't want to
+          // navigate to question 0 again because navigateToQuestion
+          // will store code for current question, thus, fail the test
+          $scope.navigateToQuestion(i);
+        }
+        expect(!$scope.autosaveOn).toBe(true);
+        var questionId = $scope.questionIds[i];
+        // There should be no code stored in local storage
+        // before autosave is triggerred.
+        expect(CodeStorageService.loadStoredCode(
```

Comment:
It might not even be a bad idea to pull it out as a separate helper function so that you can comment it thoroughly without impacting the file's code flow.